File: 백준/Gold/18116. 로봇 조립/로봇 조립.py
import sys
input = lambda: sys.stdin.readline().strip()

MAX_VALUE = 10**6+1


parent=list(range(MAX_VALUE))
cnt_list = [1]*MAX_VALUE

def find(x):
    if parent[x] != x:
        parent[x] = find(parent[x])
    return parent[x]
    
        
    

def union(a,b):
    ra, rb = find(a), find(b)
    if ra==rb:
        return
    # print('ra,rb',ra,rb)
    min_idx, max_idx = min(ra,rb), max(ra,rb)
    
    cnt_list[min_idx] += cnt_list[max_idx]
    cnt_list[max_idx] = 0
    
    parent[max_idx] = min_idx
    
    # print('parent',parent)
    # print('cnt',cnt_list)
    find(a), find(b)
    
for _ in range(int(input())):
    cmd = input().split()
    if cmd[0] == 'I':
        union(int(cmd[1]), int(cmd[2]))
    else:
        print(cnt_list[find(int(cmd[1]))])
        
    # print('parent',parent)
    # print('cnt',cnt_list)

File: 백준/Gold/18116. 로봇 조립/로봇 조립.java
import java.io.*;
import java.util.*;

public class Main {
    static final int MAX_VALUE = 1000001;
    static int[] parent = new int[MAX_VALUE];
    static int[] cntList = new int[MAX_VALUE];

    static int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    static void union(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return;
        
        int minIdx = Math.min(ra, rb);
        int maxIdx = Math.max(ra, rb);
        
        cntList[minIdx] += cntList[maxIdx];
        cntList[maxIdx] = 0;
        
        parent[maxIdx] = minIdx;
        find(a);
        find(b);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        
        int N = Integer.parseInt(br.readLine());
        
        for (int i = 0; i < MAX_VALUE; i++) {
            parent[i] = i;
            cntList[i] = 1;
        }
        
        StringBuilder sb = new StringBuilder();
        
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            String cmd = st.nextToken();
            
            if (cmd.equals("I")) {
                int a = Integer.parseInt(st.nextToken());
                int b = Integer.parseInt(st.nextToken());
                union(a, b);
            } else {
                int c = Integer.parseInt(st.nextToken());
                sb.append(cntList[find(c)]).append("\n");
            }
        }
        System.out.print(sb);
    }
}